### PR TITLE
feat(billing): meter service, extras balance, weekly reset (PR-N2)

### PR DESCRIPTION
**Pull request by PierreBrisorgueil**
## Summary

PR-N2 of the Compute Pricing layer (#3533). Builds the attribution + extras balance
services on top of PR-N1 plan versioning. All new code paths are gated behind
`config.billing.meterMode = false` (backward compat absolute — non-compute downstreams
are entirely unaffected).

### New files

- `models/billing.extraBalance.model.mongoose.js` — ledger-based extra balance doc
  (topup / debit / refund / expiration / adjustment). Idempotency indexes on
  `ledger.stripeSessionId`, `ledger.historyId`, `ledger.expiresAt`.
- `models/billing.extraBalance.schema.js` — Zod mirror with full constraint symmetry.
- `repositories/billing.extraBalance.repository.js` — atomic Mongo ops:
  `getOrCreate`, `creditPack` (idempotent on stripeSessionId), `debit` (guarded by
  cachedBalance >= amount AND refId dedup), `addExpirationEntries` (idempotent sweep),
  `getBalance` (cheap read).
- `services/billing.extra.service.js` — `creditPack`, `debit`, `expireOldEntries`,
  `refundPartial` (proportional, balance may go negative = economic reflection),
  `listLedger` (paginated, newest-first).
- `services/billing.meter.service.js` — `unitsFromCosts` (frozen plan ratio lookup,
  dollarsToUnitRatio, METER_RUN_BASE floor), `attribute` (idempotent on history._id).
- `services/billing.reset.service.js` — `resetWeek` (archive-then-upsert, E11000-safe),
  `resetAllDue` (sweeps active subs by currentPeriodStart), `isoWeekKey` helper.

### Modified files

- `repositories/billing.usage.repository.js` — add `findByWeek` (lean read),
  `incrementMeter` (atomic upsert with $ne idempotency key, E11000 retry, $setOnInsert
  snapshot).
- `services/billing.usage.service.js` — add `currentWeekKey` (ISO 8601), `incrementMeter`
  (full flow: quota snapshot, overflow→extras, 80%/100% threshold events deduped per
  cycle), `getMeter`.
- `services/billing.plan.service.js` — add `bumpVersionWithRetry` (exp backoff
  100/300/900ms on E11000, max 3 attempts).

### Architecture decisions

- **Idempotency via atomic filter**: both `creditPack` and `debit` use `$ne` guards
  inside the `findOneAndUpdate` filter so the check + update is a single round-trip
  (no TOCTOU). No separate "does it exist?" read for the happy path.
- **Threshold dedup**: `alertedAt80`/`alertedAt100` are set atomically via `updateOne`
  with `{ alertedAt80: null }` filter so only one pod wins the event emit per cycle.
- **resetWeek race safety**: `$ne weekKey` archive pass + E11000 retry on upsert
  handles concurrent pod resets gracefully.
- **refundPartial negative balance**: if units already consumed, balance goes negative —
  this is the correct economic reflection and matches the design intent.

### Tests

79 new tests (739 total, baseline 660). All edge cases from the spec:
- creditPack same stripeSessionId 2× → 1 effect
- debit > balance → applied: false
- debit same refId 2× → second no-op
- incrementMeter same historyId 2× → replay (no-op)
- incrementMeter overflow → extrasConsumed correct
- threshold 80/100 deduped per cycle
- expireOldEntries idempotent
- refundPartial consumed balance → negative (applied: true)
- bumpVersionWithRetry E11000 retry then success

## Test plan

- [x] `npm run lint` — clean
- [x] `npm run test:unit` — 739 passing
- [ ] PR review + CodeRabbit pass
- [ ] Downstream trawl_node integration (T-N2) after merge

Refs #3533

## Release Notes

* **New Features**
  * Added prepaid balance system with support for credit packs, debits, refunds, and automatic expiration.
  * Implemented weekly meter-based billing with automatic periodic resets.
  * Enhanced usage tracking with per-feature consumption breakdown.
  * Added threshold alerts at 80% and 100% of quota utilization.

### CODE REVIEW

File: modules/billing/models/billing.extraBalance.model.mongoose.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+  {
+    kind: {
+      type: String,
+      enum: ['topup', 'debit', 'refund', 'expiration', 'adjustment'],
+      required: true,
+    },
+    /**
+     * Signed amount in meter units.
+     * Positive for topup/refund/adjustment (add to balance).
+     * Negative for debit/expiration (subtract from balance).
```
[the remark this answers is not in the document]
Comment (PierreBrisorgueil) [reply modules/billing/models/billing.extraBalance.model.mongoose.js]: Fixed in f2004b5b: Updated the JSDoc in `billing.extraBalance.model.mongoose.js` and `billing.extraBalance.schema.js` to correctly document that `refund` entries carry a **negative** amount (clawback semantics). The model schema now reads: _'Positive for topup/adjustment; negative for debit/expiration/refund.'_

File: modules/billing/repositories/billing.extraBalance.repository.js [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+ * @param {string} orgId - The organization ObjectId (string).
+ * @returns {Promise<Object>} The ExtraBalance document.
+ */
+// biome-ignore lint/correctness/useQwikValidLexicalScope: false positive — Node.js repository, not Qwik
+const getOrCreate = (orgId) =>
+  BillingExtraBalance().findOneAndUpdate(
+    { organization: orgId },
+    { $setOnInsert: { organization: orgId, ledger: [], cachedBalance: 0, cachedBalanceAt: new Date() } },
+    { upsert: true, returnDocument: 'after', runValidators: true },
+  );
```
[the remark this answers is not in the document]
Comment (PierreBrisorgueil) [reply modules/billing/repositories/billing.extraBalance.repository.js:39]: Fixed in f2004b5b: Added `isValidOrgId` helper (wraps `mongoose.Types.ObjectId.isValid`) to all BillingExtraBalance repository methods (getOrCreate, creditPack, debit, addExpirationEntries, refundPartial, getBalance). Malformed orgId now returns a safe null/0 instead of letting Mongoose throw a CastError → 500.

File: modules/billing/repositories/billing.usage.repository.js [hunk 3]
```diff
@@ -65,8 +65,95 @@ const reset = (organizationId, month) => {
...
+        consumedHistoryIds: { $ne: idempotencyKey },
+      },
+      {
+        $inc: incPayload,
+        $push: { consumedHistoryIds: idempotencyKey },
+        $setOnInsert: {
+          organizationId,
+          weekKey,
+          month: baseSnapshot?.month ?? weekKey.slice(0, 7),
+          meterQuota: baseSnapshot?.meterQuota ?? 0,
```
[the remark this answers is not in the document]
Comment (PierreBrisorgueil) [reply modules/billing/repositories/billing.usage.repository.js:132]: Fixed in f2004b5b: Replaced the invalid fallback `weekKey.slice(0, 7)` (which produced `'YYYY-W'`) with an inline IIFE that derives a proper `YYYY-MM` string from `new Date()`. The upstream service always passes `baseSnapshot.month` so the fallback is a safety-net only, but it now produces a valid value.

File: modules/billing/repositories/billing.usage.repository.js [hunk 4]
```diff
@@ -65,8 +65,95 @@ const reset = (organizationId, month) => {
...
+          planVersion: baseSnapshot?.planVersion ?? null,
+          resetAt: baseSnapshot?.resetAt ?? null,
```
[the remark this answers is not in the document]
Comment (PierreBrisorgueil) [reply modules/billing/repositories/billing.usage.repository.js:134]: Fixed in f2004b5b: Added `alertedAt80: null`, `alertedAt100: null`, and `meterBreakdown: {}` to the `$setOnInsert` block in `incrementMeter`. New documents now have these fields initialized, preventing false threshold re-emissions on first-insert docs where the fields were absent vs null.

File: modules/billing/services/billing.usage.service.js [hunk 5]
```diff
@@ -1,7 +1,23 @@
...
+ */
+// biome-ignore lint/correctness/useQwikValidLexicalScope: false positive — Node.js service, not Qwik
+const getBillingEvents = async () => {
+  try {
+    const { default: billingEvents } = await import('../lib/events.js');
+    return billingEvents;
+  } catch {
+    return null;
+  }
+};
```
[the remark this answers is not in the document]
Comment (PierreBrisorgueil) [reply modules/billing/services/billing.usage.service.js:24]: Fixed in f2004b5b: Added a full `@function getBillingEvents` JSDoc block with `@async` and `@returns {Promise<EventEmitter|null>}` annotation describing the null-on-import-failure path.

File: modules/billing/services/billing.meter.service.js [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+ *              for the given organization. If the plan quota is exceeded, falls back
+ *              to BillingExtraService.debit. If extras are also exhausted, throws
+ *              MeterQuotaExhausted.
+ *
+ *              Idempotent on history._id — a second call with the same history object
+ *              is a no-op (replay protection via consumedHistoryIds).
+ *
+ * @param {Object} history - History-like object with _id, costs, planId, planVersion fields.
+ * @param {string} organizationId - The organization ObjectId (string).
+ * @returns {Promise<{applied: boolean, meterUsed: number, extrasConsumed: number}>}
```
[the remark this answers is not in the document]
Comment (PierreBrisorgueil) [reply modules/billing/services/billing.meter.service.js:71]: Fixed in f2004b5b: Updated `attribute()` JSDoc — removed the incorrect claim that it throws `MeterQuotaExhausted`. The function is best-effort: when extras are exhausted the debit returns `applied=false` and `extrasConsumed=0` is returned to the caller without throwing.

File: modules/billing/services/billing.usage.service.js [hunk 7]
```diff
@@ -41,8 +74,158 @@ const get = async (organizationId) => {
...
+        });
+      }
+    } else if (pct >= 0.8 && !updatedDoc.alertedAt80) {
+      alertCrossed = '80';
+      try {
+        const { default: mongoose } = await import('mongoose');
+        await mongoose.model('BillingUsage').updateOne(
+          { _id: updatedDoc._id, alertedAt80: null },
+          { $set: { alertedAt80: new Date() } },
+        );
```
[the remark this answers is not in the document]
Comment (PierreBrisorgueil) [reply modules/billing/services/billing.usage.service.js]: Addressed in 21a955f3: the threshold-marking update was moved to `billing.usage.repository.js` as `markThreshold(docId, field)`. The service now calls `UsageRepository.markThreshold(...)` — no direct mongoose import in the service.

File: modules/billing/tests/billing.reset.service.unit.tests.js [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+    }));
+
+    jest.unstable_mockModule('mongoose', () => ({
+      default: {
+        model: jest.fn((name) => {
+          if (name === 'BillingUsage') return mockMongooseUsage;
+          if (name === 'BillingSubscription') return mockMongooseSubscription;
+          return {};
+        }),
+      },
```
[the remark this answers is not in the document]
Comment (PierreBrisorgueil) [reply modules/billing/tests/billing.reset.service.unit.tests.js]: Addressed in 21a955f3: test now mocks the correct model name `BillingSubscription` → `Subscription` and the org field `organizationId` → `organization`, matching the actual Mongoose model registration.

File: modules/billing/services/billing.reset.service.js [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+  // biome-ignore lint/correctness/useQwikValidLexicalScope: false positive — dynamic model access
+  const { default: mongoose } = await import('mongoose');
+  await mongoose.model('BillingUsage').updateMany(
+    {
+      organizationId: orgId,
+      weekKey: { $ne: newWeekKey },
+      archivedAt: { $exists: false },
+    },
+    { $set: { archivedAt: now } },
+  );
```
[the remark this answers is not in the document]
Comment (PierreBrisorgueil) [reply modules/billing/services/billing.reset.service.js]: Addressed in 21a955f3: archive + upsert logic was extracted to `BillingUsageRepository.archiveOtherWeeks` and `BillingUsageRepository.upsertWeekSnapshot`. The service `resetWeek` is now orchestration-only — no mongoose import.

File: modules/billing/services/billing.reset.service.js [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+  const oneWeekAgo = new Date(now.getTime() - 7 * 24 * 60 * 60 * 1000);
+
+  // Find subscriptions whose period started within the last week
+  const subs = await Subscription.find(
+    {
+      status: { $in: ['active', 'trialing'] },
+      currentPeriodStart: { $gte: oneWeekAgo, $lte: now },
+    },
+    { organizationId: 1, currentPeriodStart: 1 },
+  ).lean();
```
[the remark this answers is not in the document]
Comment (PierreBrisorgueil) [reply modules/billing/services/billing.reset.service.js]: Addressed in 21a955f3: `resetAllDue` now uses `BillingSubscriptionRepository.findAllDueForReset(from, to)` which queries `mongoose.model('Subscription')` with the correct field `organization`. The direct mongoose call was removed from the service.

File: modules/billing/services/billing.extra.service.js [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+    {
+      organization: orgId,
+      'ledger.refId': { $ne: refundRefId },
+    },
+    {
+      $push: { ledger: refundEntry },
+      $inc: { cachedBalance: -refundUnits },
+      $set: { cachedBalanceAt: new Date() },
+    },
+    { returnDocument: 'after' },
```
[the remark this answers is not in the document]
Comment (PierreBrisorgueil) [reply modules/billing/services/billing.extra.service.js]: Addressed in 21a955f3: `refundPartial` in the service now delegates to `BillingExtraBalanceRepository.refundPartial(orgId, stripeSessionId, refundUnits, refId)` — no mongoose import in the service layer.

File: modules/billing/models/billing.extraBalance.schema.js [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+/**
+ * Single ledger entry schema.
+ */
+const LedgerEntry = z.object({
+  _id: z.string().trim().regex(objectIdRegex, '_id must be a valid ObjectId').optional(),
+  kind: LedgerKind,
+  /**
+   * Signed amount in meter units.
+   * Positive for topup/refund/adjustment; negative for debit/expiration.
```
[the remark this answers is not in the document]
Comment (PierreBrisorgueil) [reply modules/billing/models/billing.extraBalance.schema.js]: Fixed in f2004b5b: Updated the JSDoc in `billing.extraBalance.model.mongoose.js` (LedgerEntrySchema `amount` comment) and `billing.extraBalance.schema.js` to accurately reflect that `refund` entries are negative clawbacks. Both files now read: _'Positive for topup/adjustment; negative for debit/expiration/refund.'_

File: modules/billing/repositories/billing.usage.repository.js [hunk 13]
```diff
@@ -65,8 +65,95 @@ const reset = (organizationId, month) => {
...
+    const doc = await BillingUsage.findOneAndUpdate(
+      {
+        organizationId,
+        weekKey,
+        consumedHistoryIds: { $ne: idempotencyKey },
+      },
+      {
+        $inc: incPayload,
+        $push: { consumedHistoryIds: idempotencyKey },
+        $setOnInsert: {
```
[the remark this answers is not in the document]
Comment (PierreBrisorgueil) [reply modules/billing/repositories/billing.usage.repository.js:122]: Acknowledged. This is a known limitation deferred to a dedicated PR (PR-N4). The current design uses `consumedHistoryIds` as a per-week idempotency set — each week document has a bounded lifetime (it is archived at weekly reset and replaced with a fresh doc). So growth is bounded to the number of attributed events in a single billing week, not unbounded over the lifetime of the subscription. A TTL-based processed-events collection is the correct long-term fix and is tracked as a TODO in the schema comment. Deferring to PR-N4 to keep this PR focused on the meter core.
